Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java

```diff
@@ -2209,6 +2210,49 @@ public SpellAbility chooseAbilityForCast(Card card, Game game, boolean nonMana)
         return card.getSpellAbility();
     }
 
+    @Override
+    public SpellAbility chooseAbilityForCast(Card card, List<SpellAbility> spellAbilities, Game game) {
+        if (gameInCheckPlayableState(game)) {
+            return null;
+        }
+        if (spellAbilities.size() < 2) {
+            return spellAbilities.stream().findFirst().orElse(null);
+        }
+
+        // TODO: add canRespond cycle?
+        if (!canRespond()) {
+            return null;
+        }
+        MageObject object = game.getObject(card.getId());
+        if (object == null) {
+            return spellAbilities.get(0);
+        }
+        String message = "Choose ability to cast" + "<br>" + object.getLogName();
+        LinkedHashMap<UUID, SpellAbility> useableAbilities = new LinkedHashMap<>(
```

Comment:
That's wrong. I don't understand why you duplicated `chooseAbilityForCast` and removed playable checks at all. It allows to select non playable abilities and will cause many bugs for users and AI. Just use old method or modify it to filter abilities by id (e.g. `List<SpellAbility> filterAbilities` and `if (filterAbilities != null) useableAbilities...stream.removeif`.

Reply:
yeah I'm gonna take another look at it

Reply:
The reason I changed it was because while the current cascade code prevented cascading into Valki, God of Lies and casting Tibalt, Cosmic Impostor, it still allowed players to cast either part of an adventure, so Violent Outburst could cascade into Flaxen Intruder and cast Welcome Home. Maybe that bug should just be fixed separately.

Reply:
If you found wrong playable bug then it must be a separate issue.

XMage logic: all checking must be on server side, users can cheating and send any data in any choose dialog (e.g. send any card id instead selectable).